### PR TITLE
feature: adding assume_centered flag in Covariance

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
# Description
Adding assume centering flag to reduce computation time and increase sklearn conformance.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: cpp/daal/src/algorithms/covariance/covariance_impl.i [hunk 1]
```diff
@@ -35,7 +35,7 @@
 #include "src/algorithms/service_error_handling.h"
 #include "src/threading/threading.h"
 #include "src/externals/service_profiler.h"
-
+#include <iostream>
```
Comment (olegkkruglov) [inline cpp/daal/src/algorithms/covariance/covariance_impl.i:38]: probably need to remove

File: cpp/oneapi/dal/algo/covariance/backend/gpu/compute_kernel_dense_impl_dpc.cpp [hunk 2]
```diff
@@ -59,11 +59,16 @@ result_t compute_kernel_dense_impl<Float>::operator()(const descriptor_t& desc,
...
+    auto assume_centered = desc.get_assume_centered();
+
     auto result = compute_result<task_t>{}.set_result_options(desc.get_result_options());
+    // pr::ndarray<Float, 1> sums;
 
     const auto data_nd = pr::table2ndarray<Float>(q_, data, alloc::device);
-
-    auto [sums, sums_event] = compute_sums(q_, data_nd);
+    // sycl::event sums_event;
+    // if (!assume_centered) {
```
Comment (olegkkruglov) [inline cpp/oneapi/dal/algo/covariance/backend/gpu/compute_kernel_dense_impl_dpc.cpp:69]: why is it commented?

File: cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp [hunk 3]
```diff
@@ -43,17 +43,25 @@ namespace pr = dal::backend::primitives;
...
 
     const std::int64_t column_count = data.get_dimension(1);
-    auto sums = pr::ndarray<Float, 1>::empty(q, { column_count }, alloc::device);
-    constexpr pr::sum<Float> binary{};
-    constexpr pr::identity<Float> unary{};
-    auto sums_event = pr::reduce_by_columns(q, data, sums, binary, unary, deps);
-    return std::make_tuple(sums, sums_event);
+    if (assume_centered) {
+        auto sums = pr::ndarray<Float, 1>::empty(q, { column_count }, alloc::device);
+        auto fill_event = pr::fill(q, sums, Float(0), deps);
```
Comment (icfaust) [inline cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp:55]: Dumb question: above in compute_kernel_dense_impl_dpc.cpp you used pr::ndarray<Float, 1>::zeros to fill with zeros, and here you start with an empty and fill with zeros. Is there a difference/ which is superior? Or are there two cases here that require different approaches?

**Review by icfaust — COMMENTED** [review]
Maybe extend the testing to run with assumed_centered False and True just to make sure that the code gets exercised (in tests/batch). I hope is wouldn't be too much work. I would wait for mathematical verification on sklearnex side (which should be pretty straighforward/quick)

**Review by icfaust — APPROVED** [review]
After discussion, with a couple tiny changes I think it is good to go.

**Review by samir-nasibli — COMMENTED** [review]
Look good to me. But I also expecting tests extension for this param enabled

File: cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp (hunk 3 above)
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp:55]: here its necessary due to the function returns event+ndarray

**Comment by icfaust** [comment]
rerun of private CI based off last commit (due to infrastructure issues): http://intel-ci.intel.com/eef7bc59-abd5-f1d9-8842-a4bf010d0e2e

File: cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp (hunk 3 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/covariance/backend/gpu/misc.hpp:55]: +1 to Ian's question. `ndarray::zeros` also returns ndarray+event.
It seems you can just do `return pr::ndarray<Float, 1>::zeros(...);`.

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp [hunk 4]
```diff
@@ -79,13 +80,41 @@ inline sycl::event compute_covariance(sycl::queue& q,
...
-                cov_ptr[gi] *= multiplier;
-            }
+            cov_ptr[gi] -= inv_n * sums_ptr[id.get_id(0)] * sums_ptr[id.get_id(1)];
+            cov_ptr[gi] *= multiplier;
+        });
+    });
+}
+
+template <typename Float>
+inline sycl::event compute_covariance_centered(sycl::queue& q,
```
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:91]: Please add the description. It is not obvious that on input `cov` is a matrix of sum of squares and on output it is the covariance matrix.
The same applies to `compute_covariance`.

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp [hunk 5]
```diff
@@ -79,13 +80,41 @@ inline sycl::event compute_covariance(sycl::queue& q,
...
+
+    Float* cov_ptr = cov.get_mutable_data();
+
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = make_range_2d(p, p);
+
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::item<2> id) {
+            const std::int64_t gi = id.get_linear_id();
+            cov_ptr[gi] *= multiplier;
```
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:117]: Wouldn't it be better to have 1d range of size p * p instead?

File: examples/oneapi/cpp/source/covariance/cov_dense_biased_batch.cpp [hunk 6]
```diff
@@ -27,7 +27,8 @@ int main(int argc, char const *argv[]) {
     const auto input = dal::read<dal::table>(dal::csv::data_source{ input_file_name });
     auto cov_desc = dal::covariance::descriptor{}
                         .set_result_options(dal::covariance::result_options::cov_matrix)
-                        .set_bias(true);
+                        .set_bias(true)
+                        .set_assume_centered(true);
```
Comment (Vika-F) [inline examples/oneapi/cpp/source/covariance/cov_dense_biased_batch.cpp:31]: Shouldn't it be `false`? I think `assume_centered` can be true only when the input data is centered.

File: cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp (hunk 5 above)
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp:117]: It makes sense! Thanks!

**Pull request merged**